File: all-pairs-shortest-paths/python/floyd_warshall.py
import copy
import sys


def floyd_warshall(W):
    n = len(W)

    D = copy.deepcopy(W)
    for k in range(n):
        for i in range(n):
            for j in range(n):
                D[i][j] = min(D[i][j], D[i][k] + D[k][j])

    return D

def print_matrix(A):
    if not A:
        return

    for row in A:
        for i in range(len(row)):
            print(f"{row[i]:3d}" if row[i] < sys.maxsize else "INF", end=", ")
        print("")

if __name__ == '__main__':
    W = [
        [0, 3, 8, sys.maxsize, -4],
        [sys.maxsize, 0, sys.maxsize, 1, 7],
        [sys.maxsize, 4, 0, sys.maxsize, sys.maxsize],
        [2, sys.maxsize, -5, 0, sys.maxsize],
        [sys.maxsize, sys.maxsize, sys.maxsize, 6, 0],
    ]

    print("--- graph ---")
    print_matrix(W)

    print("--- Floyd-Warshall algorithm ---")
    D = floyd_warshall(W)

    print("--- result ---")
    print_matrix(D)

File: all-pairs-shortest-paths/golang/floyd_warshall.go
package main

import (
	"fmt"
	"math"
)

func floydWarshall(W [][]int) [][]int {
	n := len(W)
	D := make([][]int, n)
	copy(D, W)

	for k := range n {
		for i := range n {
			for j := range n {
				D[i][j] = min(D[i][j], D[i][k]+D[k][j])
			}
		}

		fmt.Printf("--- k = %d ---\n", k)
		printMatrix(D)
	}

	return D
}

func printMatrix(W [][]int) {
	for i := range W {
		for j := range W[i] {
			if W[i][j] != math.MaxInt8 {
				fmt.Printf("%3d", W[i][j])
			} else {
				fmt.Print("  -")
			}
		}

		fmt.Println("")
	}
}

func main() {
	W := [][]int{
		{0, 3, 8, math.MaxInt8, -4},
		{math.MaxInt8, 0, math.MaxInt8, 1, 7},
		{math.MaxInt8, 4, 0, math.MaxInt8, math.MaxInt8},
		{2, math.MaxInt8, -5, 0, math.MaxInt8},
		{math.MaxInt8, math.MaxInt8, math.MaxInt8, 6, 0},
	}

	fmt.Println("--- graph ---")
	printMatrix(W)

	fmt.Println("--- Flyod-Warshall algorithm ---")
	D := floydWarshall(W)

	fmt.Println("--- result ---")
	printMatrix(D)
}
